pkg/proxy/ipvs/netlink_linux.go

```diff
@@ -123,72 +124,67 @@ func (h *netlinkHandle) ListBindAddress(devName string) ([]string, error) {
-// 10.0.0.1 dev kube-ipvs0  scope host  src 10.0.0.1
-// 10.0.0.10 dev kube-ipvs0  scope host  src 10.0.0.10
-// 10.0.0.252 dev kube-ipvs0  scope host  src 10.0.0.252
-// 100.106.89.164 dev eth0  scope host  src 100.106.89.164
-// 127.0.0.0/8 dev lo  scope host  src 127.0.0.1
-// 127.0.0.1 dev lo  scope host  src 127.0.0.1
-// 172.17.0.1 dev docker0  scope host  src 172.17.0.1
-// 192.168.122.1 dev virbr0  scope host  src 192.168.122.1
-// Then cut the unique src IP fields,
-// --> result set: [10.0.0.1, 10.0.0.10, 10.0.0.252, 100.106.89.164, 127.0.0.1, 172.17.0.1, 192.168.122.1]
-
-// If dev is specified, it's equivalent to exec:
-// $ ip route show table local type local proto kernel dev kube-ipvs0
-// 10.0.0.1  scope host  src 10.0.0.1
-// 10.0.0.10  scope host  src 10.0.0.10
-// Then cut the unique src IP fields,
-// --> result set: [10.0.0.1, 10.0.0.10]
-
-// If filterDev is specified, the result will discard route of specified device and cut src from other routes.
+// GetLocalAddresses return addresses for interfaces in two ways;
+//  1. dev != "" && filterDev == ""; Return addresses for `dev`
+//  2. dev == "" && filterDev != ""; Return addresses for all interfaces *except* filterDev
+// Other combinations of dev and filterDev are forbidden.
+// IPv6 link-local addresses are excluded.
 func (h *netlinkHandle) GetLocalAddresses(dev, filterDev string) (sets.String, error) {
-	chosenLinkIndex, filterLinkIndex := -1, -1
 	if dev != "" {
-		link, err := h.LinkByName(dev)
-		if err != nil {
-			return nil, fmt.Errorf("error get device %s, err: %v", dev, err)
+		if filterDev != "" {
+			return nil, fmt.Errorf("Invalid argument")
 		}
-		chosenLinkIndex = link.Attrs().Index
-	} else if filterDev != "" {
-		link, err := h.LinkByName(filterDev)
+		ifi, err := net.InterfaceByName(dev)
 		if err != nil {
-			return nil, fmt.Errorf("error get filter device %s, err: %v", filterDev, err)
+			return nil, fmt.Errorf("Could not get interface %s: %v", dev, err)
 		}
-		filterLinkIndex = link.Attrs().Index
-	}
-
-	routeFilter := &netlink.Route{
-		Table:    unix.RT_TABLE_LOCAL,
-		Type:     unix.RTN_LOCAL,
-		Protocol: unix.RTPROT_KERNEL,
+		ips := sets.NewString()
+		h.appendAddresses(ips, ifi)
+		return ips, nil
 	}
-	filterMask := netlink.RT_FILTER_TABLE | netlink.RT_FILTER_TYPE | netlink.RT_FILTER_PROTOCOL
 
-	// find chosen device
-	if chosenLinkIndex != -1 {
-		routeFilter.LinkIndex = chosenLinkIndex
-		filterMask |= netlink.RT_FILTER_OIF
+	if filterDev == "" {
+		return nil, fmt.Errorf("Invalid argument")
 	}
-	routes, err := h.RouteListFiltered(netlink.FAMILY_ALL, routeFilter, filterMask)
+	ifaces, err := net.Interfaces()
 	if err != nil {
-		return nil, fmt.Errorf("error list route table, err: %v", err)
+		return nil, fmt.Errorf("Could not get interfaces: %v", err)
 	}
-	res := sets.NewString()
-	for _, route := range routes {
-		if route.LinkIndex == filterLinkIndex {
+	ips := sets.NewString()
+	for _, i := range ifaces {
```

Comment:
I always struggle with this , but I think that you have to capture this variable since you are passing it as reference
https://github.com/golang/go/wiki/CommonMistakes#using-reference-to-loop-iterator-variable
```
i := i
```

Reply:
I think it's ok in this case. The interface pointer &i is only used within the loop to get the addresses. It's not stored anywhere.

Reply:
yeah, but `h.appendAddresses(ips, &i)`  will only use the last value, check this example from the link I pasted
```
While this example might look a bit obvious, the same unexpected behavior could be more hidden in some other cases. For example, the loop variable can be an array and the reference can be a slice:
func main() {
	var out [][]int
	for _, i := range [][1]int{{1}, {2}, {3}} {
		out = append(out, i[:])
	}
	fmt.Println("Values:", out)
}
Output:

Values: [[3] [3] [3]]
```

Reply:
But in that example you have actually stored an address in `out`. In `h.appendAddresses(ips, &i)` the pointer is just used once.

Reply:
And I made some printouts when testing and addresses from all interfaces are checked and added.

Reply:
I told you that I always struggle with it XD, if you've tested it has to be ok